src/common/sendpacket.c

```diff
@@ -1333,3 +1381,159 @@ sendpacket_abort(sendpacket_t *sp)
 
     sp->abort = true;
 }
+#ifdef HAVE_LIBXDP
+static struct xsk_socket_info *xsk_configure_socket(struct xsk_umem_info *umem, struct xsk_socket_config* cfg, int queue_id, const char *device)
+{
+	struct xsk_socket_info *xsk;
+	struct xsk_ring_cons *rxr = NULL;
+	struct xsk_ring_prod *txr;
+	int ret;
+	xsk = (struct xsk_socket_info*)safe_malloc(sizeof(struct xsk_socket_info));
+	xsk->umem = umem;
+	ret = xsk_socket__create(&xsk->xsk, device, queue_id, umem->umem, rxr, &xsk->tx, cfg);
+	if (ret){
+        return NULL;
+    }
+
+	xsk->app_stats.rx_empty_polls = 0;
+	xsk->app_stats.fill_fail_polls = 0;
+	xsk->app_stats.copy_tx_sendtos = 0;
+	xsk->app_stats.tx_wakeup_sendtos = 0;
+	xsk->app_stats.opt_polls = 0;
+	xsk->app_stats.prev_rx_empty_polls = 0;
+	xsk->app_stats.prev_fill_fail_polls = 0;
+	xsk->app_stats.prev_copy_tx_sendtos = 0;
+	xsk->app_stats.prev_tx_wakeup_sendtos = 0;
+	xsk->app_stats.prev_opt_polls = 0;
+
+	return xsk;
+}
+
+static sendpacket_t * sendpacket_open_xdp(const char *device, char *errbuf){
+    sendpacket_t *sp;
+
+    assert(device);
+    assert(errbuf);
+
+    int nb_of_frames = 131072; // 2^17
```

Comment:
nb_of_frames is very high IMO. Can you show me that example program?
nb_of_frames * frame_size = 536870912 (536 megabytes) which is a high number.

I think the defaults will be fine. TX + RX = 2048 + 2048 = 4096.
So he allocated umem space will be 4096 * 4096 = 16777216 (16 megabytes)

Reply:
I decreased the value of nb_of_frames to 4096, unfortunately it is a bit difficult to connect UMEM to batch size, since batch size can be set by the user but nb_of_frames * frame_size should be power of two which needs additional logic, furthermore I need to modify the interface of a function which is quite widespread in the code base. Due to the shortage of time until thesis submission I would postpone this improvement. However in the documentation I'm going to mention the value of maximum batch size